// gulp-task-builder/src/builder.js
'use strict';

var gulp = require('gulp');
var pump = require('pump');
var plugins = require('./plugins');
var regexStrip = require('./regex');

var config = {
    src: './src/',
    dest: './bin/',
    temp: './temp/'
};

var defaultTaskDependencies = [];

module.exports = {
    plugins: plugins,
    config: config,
    task: createTask,
    defaultTask: defaultTask
};

function defaultTask() {
    createTask('default').depends(defaultTaskDependencies).src('./').pump();
}

function createTask(name) {

    if (!name) {
        name = Math.random().toString(36).substring(7);
    }

    defaultTaskDependencies.push(name);

    var dependencies = [];
    var tasks = [];

    var builder = {
        config: config,
        subTask: subTask
    };

    builder.depends = function (deps) {
        dependencies = deps;
        return builder;
    };

    builder.pump = function () {
        return gulp.task(name, dependencies, pumpCallback);
    };

    builder.src = function (path) {
        return subTask(gulp.src(srcFromPath(path)));
    };

    builder.temp = function () {
        return subTask(gulp.dest(config.temp)).pump();
    };

    builder.dest = function (path) {
        return subTask(gulp.dest(path ? path : config.dest)).pump();
    };

    builder.regex = function (regex, name) {
        return subTask(regexStrip(regex, name));
    };

    return plugins.extend(builder);

    function subTask(task) {
        tasks.push(task);
        return builder;
    }

    function pumpCallback(callback) {
        pump(tasks, callback);
    }
}

function srcFromPath (path) {
    if (typeof path === "string") {
        return config.src + (path || '');
    } else {
        return path.map(function (src) {
            return config.src + src;
        });
    }
}
